src/dmclock_server.h

```diff
@@ -199,15 +208,14 @@ namespace crimson {
       static double tag_calc(const Time time,
 			     double prev,
 			     double increment,
-			     uint32_t dist_req_val,
-			     bool extreme_is_high) {
+			     uint64_t dist_req_val,
+			     bool extreme_is_high,
+			     uint64_t cost) {
 	if (0.0 == increment) {
 	  return extreme_is_high ? max_tag : min_tag;
 	} else {
-	  if (0 != dist_req_val) {
-	    increment *= dist_req_val;
-	  }
-	  return std::max(time, prev + increment);
+	  double tag_increment = increment * (dist_req_val + cost);
+	  return std::max(time, prev + tag_increment);
```

Comment:
I'm not digging the code in depth yet. Sorry about that.
I just looked the code roughly and I have a question about the manner dealing the cost.
In the code, the cost is usually added to rho or delta value. However, I think cost is somewhat needed to be multiplied. Because cost is a relative value. So it should represent how much times consume the resource than the unit request(e.g. 8KB write request.)
In the dmclock paper, author said "So, for tagging purposes, a single request of IO size S is treated as equivalent to: (1+S/(Tm×Bpeak)) IO requests"
I understood that sentence as (1+S/(Tm×Bpeak)) is something like a cost and it should be multiplied with tag incremental value to deal the IO request as multiple IO requests.

Reply:
yeah, i feel the same. cost is quite the same as the number of unit requests. and both \rho_{i} and \delta_{i} denote the number of served requests. if this certain request is treated a multiple unit requests, we should certainly use something like
![ql_b2f61a2495fb7623243337d1135b7ea0_l3](https://user-images.githubusercontent.com/59071/35683457-f625cbec-079e-11e8-8401-f2b2fcdc340f.png)

Reply:
Thank you both for your detailed look into this. This is the type of discussion I was hoping we'd have!

So cost is calculated on the server side by a formula as discussed in the [ceph-devel discussion](https://marc.info/?l=ceph-devel&m=151362386931501&w=2). So yes, on a hard disk, it would have to incorporate seek times and read/write times.

One thing to keep in mind is that cost, rho, and delta are all in the same units. On the client side, when the response comes back, delta and possibly rho (if the request was handled during the reservation phase) are incremented by the cost value. See dmclock_client.h:dmclock::SimpleTracker::resp_update.

So delta (or rho) represents costs of services already received, and cost represents cost of services being scheduled to be received. Because they are in the same unit, they should not be multiplied but instead be added. It's not clear to me what ![cost_squared](http://www.sciweavers.org%2Ftex2img.php%3Feq%3Dcost%255E2%26bc%3DWhite%26fc%3DBlack%26im%3Djpg%26fs%3D12%26ff%3Darev%26edit%3D0) units would represent.

Reply:
Cost doesn't include unit concept. It's relative value like weight or ratio value that don't have any unit info.
Therefore, we need another unit concept that represents an abstract amount of resources that would be needed to process a unit request(e.g 8KB write) so that a cost can be multiplied with this unit and the result can have a meaning in the real world. This means a cost for a specific request will be represented as a ratio to the unit cost(e.g. cost for an 8KB write).
Rho & delta is also a kind of ratio value and these don't have a unit also.
So, we can multiply these values with each other multiple times.

Reply:
When I reproduce the experiment from the mclock paper, I find the behavior of the dmclock with cost integrated is not the same as expected (when `cost`=1). The reason for that is due to line 244: `double tag_increment = increment * (dist_req_val + cost)`. My understanding is that cost should replace the `delta` and `rho` (or multiply it). In `dmclock_client.h`, we already added the cost into `delta` and `rho`, so here at line 244, we should multiply `increment` with `cost` directly.

Reply:
@yzhan298 i think you mean line 224, i.e. https://github.com/ceph/dmclock/pull/46/files#diff-9674a65a2e09d3da2ab1320a086feed7R224 right?

Reply:
@tchaikov yes, I have updated the comment. Thanks

Reply:
I'd like to update everyone with some conversations that @yzhan298 and I had. @yzhan298's suggested fix cannot be correct, because it would mean that rho and delta would have no impact on tag calculation, and that cannot be correct. Once we eliminate delta and rho from tag calculation we no longer have "dmclock" but instead "mclock". The modification did produce results from the simulation more in line with expectations. But that should serve as a reminder for the importance of having a theoretical grounding for changes.

But working through that with @yzhan298 made me question the calculation of delta and rho, and that's done in the client-side trackers. So I re-evaluated that and since we're adding cost in on the server side, we no longer have to insure delta and rho are at least 1. So I modified the OrigTracker to get rid of the "1 +" and @yzhan298 was kind enough to test it out, and it produced the expected results.

In the past we'd migrated from the OrigTracker to the BorrowingTracker. The OrigTracker was written to implement the algorithm in the dmclock paper. It subtracts out responses it gets from the same server in the calculation of delta and rho.

But it seems like the new OrigTracker is the way to go.

Reply:
@TaewoongKim You are absolutely correct that _cost_ is a scalar. That is why when we discussed the addition of cost, we decided that we should re-think how we conceptualize how we think about reservation and limit. Currently they are in the unit of ops per second. But it would likely be more appropriate to think of them as _cost units per second_. So a simple/small operation might be the foundation as one cost unit. Operations that require more data to be written/read would represent multiple cost units; the same would be the case for operations that ultimately turn into multiple sub-operations.

We are even contemplating using _cost units per time unit_ that is not seconds in order to make sure reservations and limits are expressed in positive integers given some issues in sending doubles in messages.

For the initial announcement and follow-up conversation, please see the ceph-devel RFC in:
https://marc.info/?l=ceph-devel&m=151362386931501&w=2

Reply:
@tchaikov Yes, rho and delta are defined to count operations. But in the transition to thinking about costs and operations of *widely* varying costs, we need to include cost in rho and delta, which this PR does. They are measures of work done for a client by *other* servers in a given time interval (time since last request to the same server). And they should include not just # of ops but combined cost of those ops. That is why in the tracker the client increments delta by cost (and rho by cost when done in the reservation phase).

So when we calculate the tag for a new request, we need to incorporate the cost of work done by other servers and the calculated cost of this operation. We can think of it as `cost_recent_past` and `cost_this` and we therefore must add them together. Multiplying them together would result in "units" of _cost_squared_, which is not meaningful.

Reply:
@TaewoongKim The part of the dmclock paper that you quote deals specifically with characteristics of a spinning hard disk. We need something more general as we're not working strictly with hard disks, but SSDs and non-volatile memory. Plus the cost of operations sent to the OSD will vary *widely*. So we need to work with a more general concept of cost, it seems.